homeassistant/components/hyperion/camera.py

```diff
@@ -0,0 +1,255 @@
+        self._device_id = get_hyperion_device_id(server_id, instance_num)
+        self._instance_name = instance_name
+        self._client = hyperion_client
+
+        self._image_cond = asyncio.Condition()
+        self._image: bytes | None = None
+
+        # The number of open streams, when zero the stream is stopped.
+        self._image_stream_clients = 0
+
+        self._client_callbacks = {
+            f"{KEY_LEDCOLORS}-{KEY_IMAGE_STREAM}-{KEY_UPDATE}": self._update_imagestream
+        }
+
+    @property
+    def unique_id(self) -> str:
+        """Return a unique id for this instance."""
+        return self._unique_id
+
+    @property
+    def name(self) -> str:
+        """Return the name of the switch."""
+        return self._name
+
+    @property
+    def is_on(self) -> bool:
+        """Return true if the camera is on."""
+        return self.available
+
+    @property
+    def available(self) -> bool:
+        """Return server availability."""
+        return bool(self._client.has_loaded_state)
+
+    @callback
+    async def _update_imagestream(self, img: dict[str, Any] | None = None) -> None:
+        """Update Hyperion components."""
+        if not img:
+            return
+        img_data = img.get(KEY_RESULT, {}).get(KEY_IMAGE)
+        if not img_data or not img_data.startswith(IMAGE_STREAM_JPG_SENTINEL):
+            return
+        async with self._image_cond:
+            try:
+                self._image = base64.b64decode(
+                    img_data[len(IMAGE_STREAM_JPG_SENTINEL) :]
+                )
+            except binascii.Error:
+                return
+            self._image_cond.notify_all()
+
+    async def _async_wait_for_camera_image(self) -> bytes | None:
+        """Return a single camera image in a stream."""
+        async with self._image_cond:
+            await self._image_cond.wait()
+            return None if not self.available else self._image
+
+    async def _start_image_streaming_for_client(self) -> bool:
+        """Start streaming for a client."""
+        if (
+            not self._image_stream_clients
+            and not await self._client.async_send_image_stream_start()
+        ):
+            return False
+
+        self._image_stream_clients += 1
+        self.is_streaming = True
+        self.async_write_ha_state()
+        return True
+
+    async def _stop_image_streaming_for_client(self) -> None:
+        """Stop streaming for a client."""
+        self._image_stream_clients -= 1
+
+        if not self._image_stream_clients:
+            await self._client.async_send_image_stream_stop()
+            self.is_streaming = False
+            self.async_write_ha_state()
+
+    async def async_camera_image(self) -> bytes | None:
+        """Return single camera image bytes."""
+        if not await self._start_image_streaming_for_client():
+            return None
+        image = await self._async_wait_for_camera_image()
+        await self._stop_image_streaming_for_client()
```

Comment:
You do `start ... do something ... stop` also in the below method. I would suggest to make it an [async context manager](https://docs.python.org/3/library/contextlib.html#contextlib.asynccontextmanager)

Reply:
Good idea!